filebeat/reader/readfile/scanner.go

```diff
@@ -0,0 +1,80 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package readfile
+
+import "github.com/elastic/beats/libbeat/common/streambuf"
+
+type lineScanner struct {
+	in         *decoderReader
+	separator  []byte
+	bufferSize int
+
+	symlen []uint8
+	buf    *streambuf.Buffer
+}
+
+func newLineScanner(in *decoderReader, separator []byte, bufferSize int) *lineScanner {
+	return &lineScanner{
+		in:         in,
+		separator:  separator,
+		bufferSize: bufferSize,
+		buf:        streambuf.New(nil),
+		symlen:     []uint8{},
+	}
+}
+
+// Scan reads from the underlying decoder reader and returns decoded lines.
+func (s *lineScanner) scan() ([]byte, int, error) {
+	idx := s.buf.Index(s.separator)
+	for !separatorFound(idx) {
+		b := make([]byte, s.bufferSize)
+		n, err := s.in.read(b)
+		s.buf.Append(b[:n])
```

Comment:
another place where support for pre-allocating and accessing the available write-buffer in streambuf would be nice. Like:

```
s.buf.Ensure(s.bufferSize)
n, err := s.in.inread(s.buf.WriteBuffer())
```

Btw. we shoud not use streambuf here. The Reset method does set the read/write pointer to 0, but the underlying slice is not reset (buffer is reused). On big line/allocation can grow the buffer and waste memory we are never able to return.